Add ability to EBS Surrogate builder to persist properties from Marketplace source image, such as product codes

dwc0011 opened the pull request:
Add option to swap the root device to an EBS surrogate and create a new AMI in addition to being able to register an AMI.  This allows billing data to be retained when creating AMI's.  

Closes #440

dwc0011 commented:
Wanted to bump this to see if we can someone has time to review this.  Thanks

lorengordon commented:
@nywilken Sorry for the ping, but just wanted to check and see if we could get a review on this pr, or an idea whether we're on an acceptable track. What I really need to know is if there's no possibility of merging this, so we can refocus on trying to find some other way to manage the workflow. Thanks!

nywilken commented:
> @nywilken Sorry for the ping, but just wanted to check and see if we could get a review on this pr, or an idea whether we're on an acceptable track. What I really need to know is if there's no possibility of merging this, so we can refocus on trying to find some other way to manage the workflow. Thanks!

@lorengordon @dwc0011 apologies I've been out sick for the last few weeks and circling back to this PR. We have plans to review and merge.

nywilken commented on builder/ebssurrogate/step_swap_volumes.go:
```diff
@@ -0,0 +1,120 @@
+	"github.com/hashicorp/packer-plugin-sdk/template/interpolate"
+)
+
+// StepSnapshotVolumes creates snapshots of the created volumes.
```
This looks to be the comment from SnapshotVolumes

```suggestion
// StepSnapshotVolumes creates snapshots of the created volumes.
```

nywilken commented on builder/ebssurrogate/builder.go:
```diff
@@ -318,6 +318,52 @@ func (b *Builder) Run(ctx context.Context, ui packersdk.Ui, hook packersdk.Hook)
 	amiDevices := b.config.AMIMappings.BuildEC2BlockDeviceMappings()
 	launchDevices := b.config.LaunchMappings.BuildEC2BlockDeviceMappings()
 
+	var imageMethod multistep.Step
+	var volumeMethod multistep.Step
```
Given the how far down the use of these steps are I find the name to be too generic. Let me think about this a little but I do wonder if we should either use a `multistep.If(b.config.RootDevice.ImageMethod == "create", StepCreateAMI{...})` or if we should just add both the register and create steps to the steps slice and in each step check the image method to determine if the step should run.

dwc0011 replied:
Let me know which route you think would be best and I will refactor it.  I attempted to model this on the existing conditional instance step.

nywilken reviewed (commented):
I took a first pass and left a few comments. I think the if statement is simple enough but I would like for the names to be a bit more explicit on the work that is being done. ImageMethod and VolumeMethod seem simple. 

If left a question around the DeRegister step being needed. Please let me know.

dwc0011 commented on builder/ebssurrogate/builder.go:
```diff
@@ -420,34 +466,15 @@ func (b *Builder) Run(ctx context.Context, ui packersdk.Ui, hook packersdk.Hook)
 			AMIName:             b.config.AMIName,
 			Regions:             b.config.AMIRegions,
 		},
-		&StepRegisterAMI{
```
I would say yes since the create image also register the image for you.  See below
 
"Note
For Amazon EBS-backed instances, CreateImage creates and registers the AMI in a single request, so you don't have to register the AMI yourself. We recommend that you always use CreateImage unless you have a specific reason to use RegisterImage."

https://docs.aws.amazon.com/cli/latest/reference/ec2/register-image.html#:~:text=For%20Amazon%20EBS%2Dbacked%20instances,an%20AMI%20at%20any%20time.

lorengordon replied:
As confirmation, you can also see that the Deregister step is part of the `ebs` builder, which also uses CreateImage.

https://github.com/hashicorp/packer-plugin-amazon/blob/7710fbc3554e464faca17de05dd9e820e062a829/builder/ebs/builder.go#L409-L415

nywilken commented:
Hi @dwc0011 @lorengordon thanks again for pushing this change forward and for you patience with the review. Speaking about this change internally we found ourselves a bit confused on the approach. To be clear, we understand why and what is happening. But we do wonder if there is a different approach on using create or register that we can take to make this clearer to not confuse users who may already be relying on ebssurrogate. 

To help push this forward @lbajolet-hashicorp and I would like to setup a call so we can chat a bit more. Would you be up for a video call this week? We are both based in EST timezone. 

If you are up for it you can send an email to <packer at hashicorp dot com>

lorengordon commented:
Sure, @nywilken, we can hop on a call. I'm on Pacific Time, @dwc0011 is Eastern Time. Though, it looks like the email address was cutoff? If it's easier, I'm in the hangops slack as well, as `@loren`.

lbajolet-hashicorp commented:
Hi @lorengordon,

Thanks for the quick update; the email address was cutoff since Markdown interpreted it as raw html because of the braces, I've encased it in code markup, you should be able to see it now.

dwc0011 commented:
I am definitely open to a call,  I am not available from 0800-1200 on Wednesday and I am out of office on Friday.  Other than those times I am available!

Thanks,
Dennis
________________________________
From: Loren Gordon ***@***.***>
Sent: Monday, February 26, 2024 11:40 AM
To: hashicorp/packer-plugin-amazon ***@***.***>
Cc: Dennis Carey ***@***.***>; Mention ***@***.***>
Subject: Re: [hashicorp/packer-plugin-amazon] Ebssurrogate swap volumes (PR #455)


Sure, @nywilken<https://github.com/nywilken>, we can hop on a call. I'm on Pacific Time, @dwc0011<https://github.com/dwc0011> is Eastern Time. Though, it looks like the email address was cutoff? If it's easier, I'm in the hangops slack as well, as @loren.

—
Reply to this email directly, view it on GitHub<https://github.com/hashicorp/packer-plugin-amazon/pull/455#issuecomment-1964603782>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AM2LCVPARALPWFD4OYULKQ3YVS3IFAVCNFSM6AAAAABCRLZK7WVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSNRUGYYDGNZYGI>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

lorengordon commented:
> Hi @lorengordon,
> 
> Thanks for the quick update; the email address was cutoff since Markdown interpreted it as raw html because of the braces, I've encased it in code markup, you should be able to see it now.

Roger that, email sent!

lorengordon commented:
@nywilken @lbajolet-hashicorp Per the offline discussion, Dennis renamed the new argument as a bool, `use_create_image` and updated the argument description. Please let us know if you need anything else!

lbajolet-hashicorp commented on builder/ebssurrogate/step_swap_volumes.go:
```diff
@@ -0,0 +1,117 @@
+	return multistep.ActionContinue
+}
+
+func (s *StepSwapVolumes) detach_volume(ctx context.Context, ec2conn *ec2.EC2, deviceName string, volumeId string) error {
```
I would suggest renaming the function here to detachVolume, mostly since camelCase is the norm in Go
```suggestion
func (s *StepSwapVolumes) detachVolume(ctx context.Context, ec2conn *ec2.EC2, deviceName string, volumeId string) error {
```

lbajolet-hashicorp reviewed (approved):
Hi @dwc0011, @lorengordon 

Looks good to me! I left a mini nitpick on a function's name, but aside from that, we should be good to go very soon.
I'll let @nywilken do a quick final review in addition to mine before we merge, and we'll follow up with a release likely next week.

Thanks for the reroll and the chat earlier, good to see this feature in the plugin, really, thanks for your efforts!

dwc0011 commented:
Thank you for your time today, look forward to this being incorporated into packer!

nywilken commented on builder/ebssurrogate/builder.go:
```diff
@@ -318,6 +318,52 @@ func (b *Builder) Run(ctx context.Context, ui packersdk.Ui, hook packersdk.Hook)
 	amiDevices := b.config.AMIMappings.BuildEC2BlockDeviceMappings()
 	launchDevices := b.config.LaunchMappings.BuildEC2BlockDeviceMappings()
 
+	var buildAmiStep multistep.Step
```
💯

nywilken commented on builder/ebssurrogate/builder.go:
```diff
@@ -318,6 +318,52 @@ func (b *Builder) Run(ctx context.Context, ui packersdk.Ui, hook packersdk.Hook)
+	var buildAmiStep multistep.Step
+	var volumeStep multistep.Step
+
+	if b.config.RootDevice.UseCreateImage {
```
By moving this attribute up one level this should become 
```go
	if b.config.UseCreateImage {
```

nywilken reviewed (changes requested):
Thanks again for the syncing and re-rolling this change. Overall it looks good. I left a couple of suggested changes to have the configuration attribute on the builder config and not within the ami_root_device config. 

I'm running the acceptance test now and will report back if I run into any issues.

lorengordon commented:
> Thanks again for the syncing and re-rolling this change. Overall it looks good. I left a couple of suggested changes to have the configuration attribute on the builder config and not within the ami_root_device config.
> 
> I'm running the acceptance test now and will report back if I run into any issues.

Dennis is out today, but will tackle the requested change next week. Or feel free to make the changes yourself if you want to close it sooner.

dwc0011 commented:
@nywilken I have addressed all the moves and fixed the tests as well.  I appreciate the feedback and pointing us in the right direction.  I am new to go and packer in general and expected change requests, hopefully they are all complete and we are set!

Thank you again!

nywilken reviewed (approved):
@dwc0011 thanks for making the updated changes and pushing this change forward. This all looks good and the added acceptance tests are all green. Cheers! 

```=== RUN   TestAccBuilder_EbssurrogateUseCreateImageTrue
2024/03/04 11:59:30 [INFO] AWS Auth provider used: "SharedCredentialsProvider"
2024/03/04 11:59:30 Found region us-west-2
2024/03/04 11:59:30 [INFO] AWS Auth provider used: "SharedCredentialsProvider"
--- PASS: TestAccBuilder_EbssurrogateUseCreateImageTrue (198.57s)
=== RUN   TestAccBuilder_EbssurrogateUseCreateImageFalse
2024/03/04 12:01:32 [INFO] AWS Auth provider used: "SharedCredentialsProvider"
2024/03/04 12:01:32 Found region us-west-2
2024/03/04 12:01:32 [INFO] AWS Auth provider used: "SharedCredentialsProvider"
--- PASS: TestAccBuilder_EbssurrogateUseCreateImageFalse (122.11s)
=== RUN   TestAccBuilder_EbssurrogateUseCreateImageOptional
2024/03/04 12:03:03 [INFO] AWS Auth provider used: "SharedCredentialsProvider"
2024/03/04 12:03:03 Found region us-west-2
2024/03/04 12:03:03 [INFO] AWS Auth provider used: "SharedCredentialsProvider"
--- PASS: TestAccBuilder_EbssurrogateUseCreateImageOptional (91.61s)
PASS
ok      github.com/hashicorp/packer-plugin-amazon/builder/ebssurrogate  413.330s
```

nywilken commented:
We release every Tuesday so this will go out tomorrow. If I can get it released today I will trigger things earlier. Thanks again for making this change and for working with @lbajolet-hashicorp and I to better understand the needed.
